### ISSUE TITLE
Setup automated accessibility testing for the accessibility site

### ISSUE THREAD

**Issue by dmundra** [CivicActions/accessibility#52]
Related to #39, we should set up automated accessibility tests against https://civicactions.github.io/accessibility/.

**Comment by mgifford** [comment]
Possibly with https://github.com/marketplace/accesslint (which is essentially axe) - https://accesslint.com/
https://axe-linter.deque.com/ https://www.deque.com/blog/introducing-axe-linter/

**Comment by dmundra** [comment]
Another alternate is the https://pa11y.org/ tools.

**Comment by dmundra** [comment]
@grugnog and @mgifford do you know if the CivicActions home site is tested for accessibility and is it simply just locally run by folks or is there some automation configured?

**Comment by mgifford** [comment]
The CA Homesite has a few known accessibility errors associated with it. This will be addressed hopefully with the rebuild.

**Comment by grugnog** [comment]
I would lean towards pa11y as it can scan the full site (not just PR content) and I feel is more extensive and configurable than just Axe (whilst still allowing hard fails on zero-positive checks and feedback or threshold approaches on other checks.

Also, I would think we should address a11y issues now - they ones I have seen looked pretty straightforward.

**Comment by dmundra** [comment]
I like pa11y and the home site has a [sitemap](https://civicactions.com/sitemap.xml) that can be easily used by the pa11y ci to run against.

**Comment by mgifford** [comment]
The new site may not have a sitemap, but there are ways to crawl that to generate it easily enough. Tools like this can help:
https://github.com/lgraubner/sitemap-generator-cli
https://github.com/raneomik/sitemap-crawler
https://github.com/yujiosaka/headless-chrome-crawler
https://github.com/knyzorg/Sitemap-Generator-Crawler

**Comment by grugnog** [comment]
@mgifford I don't believe the tech stack is changing (at least, I haven't heard about this) with the new site. I don't think the sitemap is exhaustive though - we should probably use the list of pages that Gatsby generates each build (at page-data/index/page-data.json) which I believe is complete.

**Comment by mgifford** [comment]
Does this contain all of the pages? I looked at [https://civicactions.com/page-data/index/page-data.json](https://civicactions.com/page-data/index/page-data.json)

But it seems there are less than what I found when I googled: **inurl:https://civicactions.com**

**Comment by dmundra** [comment]
Ya looks like that JSON URL is reporting on the root page / (e.g. https://civicactions.com/page-data/values/page-data.json is for /values page). There might be a JSON URL for all pages but I am not familiar with Gatsby.

**Comment by grugnog** [comment]
Ahh, yeah there is a data file for each page, so we would need to walk the static files though. I am sure we can extract a complete list during the build process though.

**Comment by grugnog** [comment]
Looks like we can add a snippet that writes paths on build https://github.com/CivicActions/civicactions.com/pull/911 (thanks to https://stackoverflow.com/a/60308166 for the copypasta!)

**Comment by dmundra** [comment]
@grugnog found the Docker pa11y image https://github.com/drydockcloud/ci-pa11y (from [slack thread](https://civicactions.slack.com/archives/CCX8C4MCY/p1570747362000300?thread_ts=1570558398.001500&cid=CCX8C4MCY) and Tom). Linking it here for reference.

**Comment by lukefretwell** [comment]
@dmundra can we change this issue related to just the a11y site?

**Comment by dmundra** [comment]
@lukefretwell yes we can do that.

**Comment by dmundra** [comment]
I got pa11y automated accessibility testing running on my fork using GitHub actions. The code is viewable in this PR https://github.com/dmundra/accessibility/pull/2 (but I will submit a PR back to this repo after the previous one #90 is merged).

Here is an example of what a successful run looks like https://github.com/dmundra/accessibility/actions/runs/556279894 and failure run https://github.com/dmundra/accessibility/actions/runs/556284779. The successful one was run just against the homepage which is clear of any errors. The failures are caused by two current issues:

* Most pages have this one which is just a bug in the layout files that can be fixed.

```
Error: Duplicate id attribute value "main-content" found on the web page.
WCAG2AA.Principle4.Guideline4_1.4_1_1.F77

<main class="usa-layout-docs usa-layout-docs__main usa-prose" id="main-content">
                    <ul class=...</main> (select with "#main-content")
```

* The VPAT HTML page has these which should be fixable.

```
Error: A title should be provided for the document, using a non-empty title element in the head section.
WCAG2AA.Principle2.Guideline2_4.2_4_2.H25.1.NoTitleEl

<head><meta content="text/html; chars...</head> (select with "html > head")
```
```
Error: The html element should have a lang or xml:lang attribute which describes the language of the document.
WCAG2AA.Principle3.Guideline3_1.3_1_1.H57.2

<html><head><meta content="text/html;...</html> (select with "HTML")
```
```
Error: This link points to a named anchor "id.ihv636" within the document, but no anchor exists with that name.
WCAG2AA.Principle2.Guideline2_4.2_4_1.G1,G123,G124.NoSuchID

<a class="c9" href="#id.ihv636">WCAG 2.0</a> (select with "html > body > table:nth-child(62) > tbody > tr:nth-child(5) > td:nth-child(2) > p > span:nth-child(2) > a")
```

I think the above issues are fixable. I also left the default configuration for pa11y-ci but we can configure it to test for section508 to AA and so on.

**Comment by dmundra** [comment]
Todos

- [ ] Adjust pa11y-ci configuration to test the accessibility site using section508 or AA or <blank>
- [ ] Maybe add pa11y-ci button to display status on the github page
- [ ] Submit PR to merge the changes to this repo.

**Comment by mgifford** [comment]
Ideally we'd be using the axe engine with Pa11y which defaults to WCAG 2.1 AA on the tests that it runs.

Looking forward to seeing this in action.

**Comment by dmundra** [comment]
@mgifford pa11y-ci runs WCAG2AA by default. When I run with the 'axe' engine we get a lot more errors on the accessibility site including 50 for the VPAT page. The results are available here https://github.com/dmundra/accessibility/actions/runs/558268784.
